### ISSUE TITLE
A question about technical details.

### ISSUE THREAD

**Issue by tanbuzheng** [joshyZhou/ASTv2#3]
Dear author! I'd like to ask about a technical detail. For the task of deraining, why did you adopt a progressive training from 128x128 to 256x256 during training, but the resolution of the test images is 512x512?

**Comment by joshyZhou** [comment]
Hi,

Thank you for your interest in our work! We adopt progressive training from 128×128 to 256×256 because the training datasets provide samples at a 256×256 resolution. During evaluation, we test on 512×512 patches to ensure a fair comparison with existing approaches such as DRSformer and Uformer, which also use this resolution.

I hope this clarifies your question. Please feel free to reach out if you have any further concerns.

Best regards,
Zhou

**Comment by tanbuzheng** [comment]
Thank you very much for your answer. The purpose of my asking the above question is to figure out whether such progressive training (from 128 to 256) is helpful in achieving better results on images of higher resolution (512)? Because for my task, training directly on 512x512 images seems to consume more computing resources. Please forgive the simplicity of my question and look forward to your guidance.

**Comment by joshyZhou** [comment]
Hi,

According to the discussion in Restormer (CVPR’22):
“The model trained on mixed-size patches via progressive learning shows enhanced performance at test time where images can be of different resolutions (a common case in image restoration).”

The rationale behind this strategy follows the idea of curriculum learning, where the model learns from simpler (smaller) samples first and then gradually adapts to more complex (larger) inputs. This helps the network generalize better to higher-resolution images, such as 512×512.

As you mentioned, directly training on large images consumes significantly more computational resources. Progressive learning alleviates this issue by reducing the batch size as the patch size increases, maintaining approximately the same training time per optimization step as fixed-size training.

I hope this explanation helps!

Best regards,
Zhou

**Comment by tanbuzheng** [comment]
Thank you very much for your valuable experience~

**Comment by tanbuzheng** [comment]
Sorry to bother you again, but I still have some questions. In adaptive sparse self-attention, why does the modulation of DSA by SSA not occur before the softmax operation? Would this also be acceptable? Additionally, is the final adopted self-attention map the sum of the dense self-attention map and the dense self-attention map modulated by SSA, as shown in the code? This seems to be somewhat different from the paper.

**Comment by joshyZhou** [comment]
Hi，
(1)The reason why we implement the modulation of DSA by SSA in the current formulation is that noisy interaction is attributed to the softmax function, which reweights the unrelated interactions with overly high scores. A similar discussion can be found in [issue3 in AST](https://github.com/joshyZhou/AST/issues/10).
(2)No, the final adopted self-attention map is the dense self-attention map modulated by SSA. The code can be traced to line 151 in [ASTv2_arch.py](https://github.com/joshyZhou/ASTv2/blob/main/basicsr/models/archs/ASTv2_arch.py).
I am not sure if you checked the code of the conference version?

**Comment by tanbuzheng** [comment]
Thank you for your patient response. I understand the first question now. Regarding the second one, I traced it to line 151 of the code `attn = attn0 * (1 + scale) + shift`. I think I might have misunderstood it before. My original understanding was that if the scaling factor was entirely derived from the sparse self-attention map, the over-sparse issue would still persist. However, if the scaling factor is `1 + scale`, this problem seems to be balanced out. So, is this the correct way to understand it?

**Comment by joshyZhou** [comment]
Hi,
According to the results reported in Table X, ASSA in this form achieves a `balanced' score at Entrocy with better performance. Based on this observation, we might say that the over-sparse issue is avoid.

**Comment by tanbuzheng** [comment]
Ok, thank you again.



---- Replied Message ----
| From | ***@***.***> |
| Date | 10/16/2025 11:32 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [joshyZhou/ASTv2] A question about technical details. (Issue #3) |
joshyZhou left a comment (joshyZhou/ASTv2#3)

Hi,
According to the results reported in Table X, ASSA in this form achieves a `balanced' score at Entrocy with better performance. Based on this observation, we might say that the over-sparse issue is avoid.

**Comment by tanbuzheng** [comment]
Hi auther, I have one more question. It's about the lines 146-147 in [ASTv2_arch.py](https://github.com/joshyZhou/ASTv2/blob/main/basicsr/models/archs/ASTv2_arch.py).  
attn1 = rearrange(attn1, 'b head L c -> b c head L')
x_att=self.att_ca(attn1)
I do not understand why "rearrange" is required to swap the order of h and c before calculating the modulation maps.

I still don't understand this issue. I'm looking forward to your reply.

**Comment by joshyZhou** [comment]
Hi,

Sorry for the late reply. The reason for this operation is that without the rearrangement, the convolution would operate along the wrong axes and fail to learn meaningful modulation. The rearrangement ensures that the channel dimension is correctly aligned for the 1x1 conv operation.

I hope this helps clarify your question.

Best regards,
Zhou

**Comment by tanbuzheng** [comment]
Thank you so much. I think I can probably understand now.
